File: code/Python/200-Number of Islands.py
# Given a 2d grid map of '1's (land) and '0's (water), count the number of islands. An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.
#
# Example 1:
# Input:
# 11110
# 11010
# 11000
# 00000
# Output: 1
#
# Example 2:
#
# Input:
# 11000
# 11000
# 00100
# 00011
# Output: 3
from collections import deque
class Solution(object):
    # def dfs(self, grid, r, c):
    #     grid[r][c] = 0
    #     nr, nc = len(grid), len(grid[0])
    #     for x, y in [(r - 1, c), (r + 1, c), (r, c - 1), (r, c + 1)]:
    #         if 0 <= x < nr and 0 <= y < nc and grid[x][y] == '1':
    #             self.dfs(grid, x, y)
    #
    # def numIslands(self, grid):
    #     """
    #     :type grid: List[List[str]]
    #     :rtype: int
    #     """
    #     nr = len(grid)
    #     if nr == 0:
    #         return 0
    #     nc = len(grid[0])
    #
    #     ans = 0
    #     for r in range(nr):
    #         for c in range(nc):
    #             if grid[r][c] == '1':
    #                 ans += 1
    #                 self.dfs(grid, r, c)
    #     return ans
    def numIslands(self, grid):
        nr = len(grid)
        if nr == 0:
            return 0
        nc = len(grid[0])
        ans = 0
        for r in range(nr):
            for c in range(nc):
                if grid[r][c] == '1':
                    ans += 1
                    grid[r][c] = '0'
                    neighbours = deque([(r, c)])
                    while neighbours:
                        row, col = neighbours.popleft()
                        for x, y in [(row, col - 1), (row, col + 1), (row - 1, col), (row + 1, col)]:
                            if 0 <= x < nr and 0 <= y < nc and grid[x][y] == '1':
                                neighbours.append((x, y))
                                grid[x][y] = "0"
        return ans

File: code/C++/200-Number of Islands.cpp
// Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.

// An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

 

// Example 1:

// Input: grid = [
//   ["1","1","1","1","0"],
//   ["1","1","0","1","0"],
//   ["1","1","0","0","0"],
//   ["0","0","0","0","0"]
// ]
// Output: 1
// Example 2:

// Input: grid = [
//   ["1","1","0","0","0"],
//   ["1","1","0","0","0"],
//   ["0","0","1","0","0"],
//   ["0","0","0","1","1"]
// ]
// Output: 3
 

// Constraints:

// m == grid.length
// n == grid[i].length
// 1 <= m, n <= 300
// grid[i][j] is '0' or '1'.


// dfs
class Solution {
private:
    int directions[4][2] = {
        {1, 0},
        {-1, 0},
        {0, -1},
        {0, 1},
    };
    void dfs(vector<vector<char>>& grid, int r, int c) {
        int m = grid.size();
        int n = grid[0].size();

        grid[r][c] = '0';
        for (int i = 0; i < 4; i ++) {
            int nr = r + directions[i][0];
            int nc = c + directions[i][1];
            if (nr >= 0 && nr < m && nc >= 0 && nc < n && grid[nr][nc] == '1') {
                dfs(grid, nr, nc);
            }
        }
    }

public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        if (!m) {
            return 0;
        }
        int n = grid[0].size();
        int num_islands = 0;
        for (int r = 0; r < m; r ++) {
            for (int c = 0; c < n; c ++){
                if (grid[r][c] == '1') {
                    num_islands ++;
                    dfs(grid, r, c);
                }
            }
        }
        return num_islands;
    }
};


// bfs


class Solution {
private:
    int directions[4][2] = {
        {1, 0},
        {-1, 0},
        {0, -1},
        {0, 1},
    };

public:
    int numIslands(vector<vector<char>>& grid) {
        int m = grid.size();
        if (!m) {
            return 0;
        }
        int n = grid[0].size();

        int numIsland = 0;
        for (int r = 0; r < m; r ++) {
            for (int c = 0; c < n; c ++) {
                if (grid[r][c] == '1') {
                    numIsland ++;
                    grid[r][c] = '0';
                    queue<pair<int, int>> neighbors;
                    neighbors.push({r, c});
                    while (!neighbors.empty()) {
                        int row = neighbors.front().first;
                        int col = neighbors.front().second;
                        neighbors.pop();
                        for (int i = 0; i < 4; i++) {
                            int nr = row + directions[i][0];
                            int nc = col + directions[i][1];
                            if (nr >= 0 && nr < m && nc >= 0 && nc < n && grid[nr][nc] == '1') {
                                neighbors.push({nr, nc});
                                grid[nr][nc] = '0';
                            }
                        }
                    }
                }
            }
        }
        return numIsland;
    }
};